tritium-core/src/main/java/com/palantir/tritium/event/CompositeInvocationEventHandler.java

```diff
@@ -32,38 +32,41 @@
 
     private static final Logger logger = LoggerFactory.getLogger(CompositeInvocationEventHandler.class);
 
-    private final List<InvocationEventHandler<InvocationContext>> handlers;
+    private final InvocationEventHandler<InvocationContext>[] handlers;
 
+    @SuppressWarnings("unchecked")
     private CompositeInvocationEventHandler(List<InvocationEventHandler<InvocationContext>> handlers) {
-        this.handlers = ImmutableList.copyOf(handlers);
+        this.handlers = checkNotNull(handlers, "handlers").toArray(new InvocationEventHandler[0]);
+        for (InvocationEventHandler<InvocationContext> handler : handlers) {
+            checkNotNull(handler, "Null handlers are not allowed");
+        }
     }
 
     public static InvocationEventHandler<InvocationContext> of(
             List<InvocationEventHandler<InvocationContext>> handlers) {
         if (handlers.isEmpty()) {
             return NoOpInvocationEventHandler.INSTANCE;
         } else if (handlers.size() == 1) {
-            return checkNotNull(handlers.iterator().next(), "at index 0");
+            return checkNotNull(handlers.iterator().next(), "Null handlers are not allowed");
```

Comment:
IntelliJ was yelling at me about this the other day to avoid the iterator allocation:

```suggestion
            return checkNotNull(handlers.get(0), "Null handlers are not allowed");
```